### PR TITLE
Fixed: Integrate with Akismet for added spam filtering #673

### LINKED ISSUE #673: Integrate with Akismet for added spam filtering
Besides our Honeypot, we could lean on Akismet to help filter more spam donations out. Here's how Contact Form 7 does it:

http://contactform7.com/spam-filtering-with-akismet/

Here's their source:
https://github.com/wp-plugins/contact-form-7/blob/ba71e5ba72764dc08c4987f76d135d1ef19bd9a7/modules/akismet.php

### PR DESCRIPTION
**Pull request by jaydeeprami**
## Description

Fix #673

- I have implemented Akismet API and detect Donor with pass First name and Email address in Akismet API. Based on its response, show error message on Frontend. 

@DevinWalker, You can review and let me know that is there any changes which I need to do here.

Please review video:
http://www.screencast.com/t/dMUDPAAyP9

## Screenshot
screenshot-givefee dev1 in-2017-09-18-13-31-24-464

## How Has This Been Tested?

- Active and Deactive Akismet API.
- Its only work with Akismet plugin active.
- It will consider any donor as a spam if it will try multiple times in a short period of time with same IP address and not allowed to place Donation,  After 8-10 min, they can able to access and place the order.

## Types of changes

- It will prevent spam donor and give the error message on the Frontend side.
- Spam detect with Akismet API same as Contact Form 7.

## Checklist:
- [x] My code is tested.
- [X] My code follows the WordPress code style.
- [X] My code follows has proper inline documentation.

### CODE REVIEW

File: includes/filters.php [hunk 1]
```diff
@@ -65,4 +65,102 @@ function give_set_settings_with_disable_prefix( $old_settings, $settings ) {
...
+ */
+function give_akismet( $spam ) {
+
+	// Bail out, If spam.
+	if ( $spam ) {
+		return $spam;
+	}
+
+	// Bail out, if Akismet key not exist.
+	if ( ! give_check_akismet_key() ) {
```
Comment (DevinWalker) [inline includes/filters.php]: Nice

File: includes/filters.php [hunk 2]
```diff
@@ -65,4 +65,102 @@ function give_set_settings_with_disable_prefix( $old_settings, $settings ) {
...
+		$response = akismet_http_post( $query_string, $akismet_api_host,
+			'/1.1/comment-check', $akismet_api_port );
+	}
+
+	// It's spam if response status is true.
+	if ( 'true' == $response[1] ) {
+		$spam = true;
+	}
+
+	return apply_filters( 'give_akismet_span_check', $spam, $args );
```
Comment (DevinWalker) [inline includes/filters.php]: Typo here `span` should be `spam` - please update

File: includes/process-donation.php [hunk 3]
```diff
@@ -286,6 +286,11 @@ function give_donation_form_validate_fields() {
 		give_set_error( 'invalid_honeypot', esc_html__( 'Honeypot field detected. Go away bad bot!', 'give' ) );
 	}
 
+	// Check spam detect.
+	if ( is_spam_donation() ) {
+		give_set_error( 'invalid_donation', __( 'Spam detected. Go away bad bot!', 'give' ) );
```
Comment (DevinWalker) [inline includes/process-donation.php]: Please update message `This donation has been flagged as spam. Please try again.`

File: includes/process-donation.php [hunk 4]
```diff
@@ -326,6 +331,25 @@ function give_donation_form_validate_fields() {
 }
 
 /**
+ * Detect spam donation.
+ *
+ * @since 1.8.15
+ *
+ * @return bool|mixed
+ */
+function is_spam_donation() {
```
Comment (DevinWalker) [inline includes/process-donation.php]: Please prefix with `give` so it would be `give_is_spam_donation`

**Comment by ravinderk** [comment]
@jaydeeprami Please add documentation to newly created filters

File: includes/filters.php [hunk 5]
```diff
@@ -65,4 +65,102 @@ function give_set_settings_with_disable_prefix( $old_settings, $settings ) {
 		update_option( 'give_settings', $settings );
 	}
 }
+
 add_action( 'update_option_give_settings', 'give_set_settings_with_disable_prefix', 10, 2 );
+
+/**
+ * Check spam through Akismet.
+ *
+ * @since 1.8.15
```
Comment (ravinderk) [inline includes/filters.php]: @jaydeeprami This issue shifted from `1.8.15` to `1.8.14`. Can you update it in this PR too?

Comment (jaydeeprami) [reply includes/filters.php]: @ravinderk Sure, I will do.

**Comment by jaydeeprami** [comment]
@DevinWalker and @ravinderk, I have done all changes and fixed gateway redirect issue. I have tested it from my end and it's working fine.

Thanks,

**Comment by DevinWalker** [comment]
@jaydeeprami please include an option under _Donations > Settings > Advanced_ to enable or disable the Akismet spam protection. By default, it should be _enabled_ if akismet is installed and configured. Otherwise, it will be disabled. 

- Field title "Akismet SPAM Protection" 
- Options "enabled" or "disabled" - enabled by default if Akismet is configured. Disabled if the plugin isn't installed or is installed but not configured. There should be checks in place when this option is saved to see if Akismet is enabled properly. If an admin attempts to save it as enabled and it isn't configured a helpful error will display saying "Please properly configure Akismet to enable SPAM protection". 
- Description: "Add a layer of SPAM protection to your donation submissions with Akismet. When enabled, donation submissions will be first sent to Akismet's API

**Comment by jaydeeprami** [comment]
Okay @DevinWalker, I will work on it. Thanks

**Comment by jaydeeprami** [comment]
@DevinWalker, I have done all changes and test again. You can review it.
Thanks,

File: includes/admin/class-admin-settings.php [hunk 6]
```diff
@@ -91,6 +91,16 @@ public static function save() {
 				die();
 			}
 
+			// Show error message if Akismet not configured and Admin try to save 'enabled' option.
+			if ( isset( $_POST['akismet_spam_protection'] )
+			     && 'enabled' === $_POST['akismet_spam_protection']
```
Comment (ravinderk) [inline includes/admin/class-admin-settings.php]: @jaydeeprami Can you use` give_is_setting_enabled`?

Comment (jaydeeprami) [reply includes/admin/class-admin-settings.php]: Sure @ravinderk

Comment (jaydeeprami) [reply includes/admin/class-admin-settings.php]: @ravinderk, I have updated it.

**Pull request merged**